### PR TITLE
Adds OIDC code workflow

### PR DESCRIPTION
**Pull request by jvanderhoof**
### Desired Outcome

This PR adds support for OIDC's code-based authorization flow. This functionality allows Conjur to use an external
OIDC Provider (Okta, Keycloak), to authenticate users and redirect them to Conjur on success. Conjur will resolve the identified user to a Conjur User ID and return and Authorization Token.

### Implemented Changes

The architecture of this feature is different from that of the other authenticators. The intention of this architecture is to simplify testing and reduce developer cognitive load.

- Authenticator Specific Constructs:
    - Authenticator Data Object - contains the configuration data specific to an authenticator type.
    - Strategy - implements the lower level details (the "how") an authenticator verifies an external identity.
    - Resolve Identity - functionality to an externally identified identity with an internal identity.
    - Client - handles the low level network specifics of verifying an identity with an external identity provider.
- Generic Authenticator Constructs:
    - Authentication Handler - a generic handler which calls the authenticator specific objects to orchestrate the authentication flow.

### Planned Future Work

- Add tests for Authentication Handler: [ONYX-22990](https://ca-il-jira.il.cyber-ark.com:8443/browse/ONYX-22990)
- Document Authenticator Architecture: [ONYX-22991](https://ca-il-jira.il.cyber-ark.com:8443/browse/ONYX-22991)

### Connected Issue/Story

[Jira Epic](https://ca-il-jira.il.cyber-ark.com:8443/browse/ONYX-17888)

### Definition of Done
*At least 1 todo must be completed in the sections below for the PR to be
merged.*

#### Changelog

- [x] The CHANGELOG has been updated, or
- [ ] This PR does not include user-facing changes and doesn't require a
  CHANGELOG update

#### Test coverage

- [x] This PR includes new unit and integration tests to go with the code
  changes, or
- [ ] The changes in this PR do not require tests

#### Documentation

- [ ] Docs (e.g. `README`s) were updated in this PR
- [ ] A follow-up issue to update official docs has been filed here: [insert issue ID]()
- [ ] This PR does not require updating any documentation

#### Behavior

- [x] This PR changes product behavior and has been reviewed by a PO, or
- [ ] These changes are part of a larger initiative that will be reviewed later, or
- [ ] No behavior was changed with this PR

#### Security

- [x] Security architect has reviewed the changes in this PR,
- [ ] These changes are part of a larger initiative with a separate security review, or
- [ ] There are no security aspects to these changes

### CODE REVIEW

**Review by mFelgate — APPROVED** [review]
lgtm (other then the error handling haha)

File: app/db/repository/authenticator_repository.rb [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+        ).all.map do |webservice|
+          load_authenticator(account: account, id: webservice.id.split(':').last, type: type)
+        end.compact
+      end
+
+      def find(type:, account:,  service_id:)
+        webservice =  @resource_repository.where(
+          Sequel.like(
+            :resource_id,
+            "#{account}:webservice:conjur/#{type}/#{service_id}%"
```
Comment (telday) [inline app/db/repository/authenticator_repository.rb:25]: It looks like this function is trying to find a specific webservice but wouldn't the `%` at the end allow it to pickup other rows from the db we don't necessarily want? Or am I missing something with this function?

File: app/domain/authentication/handler/authentication_handler.rb [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+            authenticator: authenticator
+          ).callback(parameters),
+          account: parameters[:account],
+          allowed_roles: @role.that_can(
+            :authenticate,
+            @resource[authenticator.resource_id]
+          ).all
+        )
+
+        # TODO: Add an error message
```
Comment (telday) [inline app/domain/authentication/handler/authentication_handler.rb:61]: Do we want to address this before merging or leave for later

File: app/domain/authentication/handler/authentication_handler.rb [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+      rescue => e
+        log_audit_failure(parameters[:account], parameters[:service_id], request_ip, @authenticator_type, e)
+        handle_error(e)
+      end
+
+      def handle_error(err)
+        @logger.warn(@authentication_error.new(err))
+        case err
+        when Errors::Authentication::Security::RoleNotAuthorizedOnResource
+          puts "Errors::Authentication::Security::RoleNotAuthorizedOnResource"
```
Comment (telday) [inline app/domain/authentication/handler/authentication_handler.rb:83]: Out of curiosity why are we printing the error class here?

We could we have a `puts err.class` before the switch statement to eliminate most of the print statements

File: app/db/repository/authenticator_repository.rb (hunk 1 above)
Comment (jvanderhoof) [reply app/db/repository/authenticator_repository.rb:25]: good catch @telday!

**Pull request merged**